lib/ansible/parsing/vault/lib.py

```diff
@@ -0,0 +1,328 @@
+
+        try:
+            this_cipher = CIPHER_MAPPING[self.cipher_name]()
+        except KeyError:
+            raise AnsibleError(f"{self.cipher_name} cipher could not be found")
+
+        # encrypt data
+        if vault_id:
+            display.vvvvv(f'Encrypting with vault_id "{to_text(vault_id)}" and vault secret {to_text(secret)}')
+        else:
+            display.vvvvv(f"Encrypting without a vault_id using vault secret {to_text(secret)}")
+
+        b_ciphertext = this_cipher.encrypt(b_plaintext, secret, salt)
+
+        # format the data for output to the file
+        return format_vaulttext_envelope(b_ciphertext, self.cipher_name, vault_id=vault_id)
+
+    def decrypt(self, vaulttext, filename=None, obj=None):
+        '''Decrypt a piece of vault encrypted data.
+
+        :arg vaulttext: a string to decrypt.  Since vault encrypted data is an
+            ascii text format this can be either a byte str or unicode string.
+        :kwarg filename: a filename that the data came from.  This is only
+            used to make better error messages in case the data cannot be
+            decrypted.
+        :returns: a byte string containing the decrypted data and the vault-id that was used
+
+        '''
+        plaintext, vault_id, vault_secret = self.decrypt_and_get_vault_id(vaulttext, filename=filename, obj=obj)
+        return plaintext
+
+    def decrypt_and_get_vault_id(self, vaulttext, filename=None, obj=None):
+        """Decrypt a piece of vault encrypted data.
+
+        :arg vaulttext: a string to decrypt.  Since vault encrypted data is an
+            ascii text format this can be either a byte str or unicode string.
+        :kwarg filename: a filename that the data came from.  This is only
+            used to make better error messages in case the data cannot be
+            decrypted.
+        :returns: a byte string containing the decrypted data and the vault-id vault-secret that was used
+
+        """
+        b_vaulttext = to_bytes(vaulttext, errors='strict', encoding='utf-8')
+
+        if self.secrets is None:
+            msg = "A vault password must be specified to decrypt data"
+            if filename:
+                msg += f" in file {to_native(filename)}"
+            raise AnsibleVaultError(msg)
+
+        if not is_encrypted(b_vaulttext):
+            msg = "input is not vault encrypted data. "
+            if filename:
+                msg += f"{to_native(filename)} is not a vault encrypted file"
+            raise AnsibleError(msg)
+
+        b_vaulttext, dummy, cipher_name, vault_id = parse_vaulttext_envelope(b_vaulttext, filename=filename)
+
+        # create the cipher object, note that the cipher used for decrypt can
+        # be different than the cipher used for encrypt
+        if cipher_name in CIPHER_WHITELIST:
+            this_cipher = CIPHER_MAPPING[cipher_name]()
+        else:
+            raise AnsibleError(f"{cipher_name} cipher could not be found")
+
```

Comment:
Please put back initializing `b_plaintext` and `vault_id_used` in the outer scope. I think removing it makes it easier to potentially introduce issues in the try/except below.